mgr/cephadm: Stop NFS service/daemon from starting automatically after reboot, cephadm to manage startup

ShwetaBhosale1 opened the pull request:
PR fixes:
1. Disable NFS daemon service, so that daemon will not start automatically after reboot. And serve loops '_check_daemons' method will start the daemon. If service is configured in active-passive mode (placement count < hosts), then the daemon will not start on rebooted node (as failed over to another node) and it will directly get removed. Daemon will be only started if they needs to be started (If same rank daemon exists then only start highest rank generation daemon).
2. Since the server loop is now starting the NFS daemons, we must ensure that daemons stopped by the user are not restarted. Added user_stopped field in DaemonDescription to track it.
3. For one NFS instance case, as fence_old_rank was happening in every loop, the down node's rank map was getting removed and daemon was getting redeployed once node is up.
4. The case where NFS is deployed with just placement hosts, when node with low rank was getting down, the node with high rank was getting redeployed with the rank of down daemon.
 

Fixes: https://tracker.ceph.com/issues/73442
Signed-off-by: Shweta Bhosale <Shweta.Bhosale1@ibm.com>



## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)

You must only issue one Jenkins command per-comment. Jenkins does not understand
comments with more than one command.
</details>

ShwetaBhosale1 committed:
mgr/cephadm: Stop NFS service/daemon from starting automatically after reboot, cephadm to manage startup

Fixes: https://tracker.ceph.com/issues/73442
Signed-off-by: Shweta Bhosale <Shweta.Bhosale1@ibm.com>

adk3798 commented on src/pybind/mgr/cephadm/schedule.py:
```diff
@@ -382,7 +383,7 @@ def expand_candidates(ls: List[DaemonPlacement], num: int) -> List[DaemonPlaceme
         to_add: List[DaemonPlacement] = []
         to_remove: List[orchestrator.DaemonDescription] = []
-        ranks: List[int] = list(range(len(candidates)))
+        ranks: List[int] = list(range(len(all_candidates)))
```
at a glance it looks like the expected difference between all_candidates and candidates for an nfs service (which I believe doesn't allow co-location currently) would just be the removal of offline hosts from the candidate lists. What's the reason we now need to consider offline hosts as well when figuring out the the full set of ranks we should check against (or if I'm off on what this accomplishes and it's something other than being about offline hosts I guess my question would be what is the purpose)?

ShwetaBhosale1 replied:
I observed that when the NFS service was deployed with placement hosts (e.g., 4 hosts, with the daemon expected to run on all 4), if a node with a rank lower than the highest (e.g., rank 0) went down, the offline host was removed from the candidates list. This resulted in the ranks being updated to [0, 1, 2], which caused the matches_rank_map check to fail for the daemon with rank 3 and that daemon was getting redeployed (removed and added) with rank 0. And when the down node came back online, that daemon was also getting redeployed. To avoid this, considering all_candidates. I don't think it will cause any issue for other services/cases. Let me know if i am missing anything.

ShwetaBhosale1 committed:
mgr/cephadm: Only start ranked daemon with highest rank gen, if many daemon exists with same rank

If the removal of an extra NFS daemon fails for any reason, that daemon should not be started, as the NFS daemon with the highest rank generation is already running.

Fixes: https://tracker.ceph.com/issues/73442
Signed-off-by: Shweta Bhosale <Shweta.Bhosale1@ibm.com>

adk3798 commented:
jenkins render docs

ShwetaBhosale1 committed:
mgr/cephadm: Fixed cache update for updating daemon user_stopped status
Fixes: https://tracker.ceph.com/issues/73442

Signed-off-by: Shweta Bhosale <Shweta.Bhosale1@ibm.com>

ShwetaBhosale1 committed:
mgr/cephadm: Fixed rank calculation logic for potential candidates when NFS daemon colocation is enabled

When colocation of NFS daemons is allowed, the existing rank calculation incorrectly derives the
rank based solely on the number of hosts. This leads to inaccurate rank assignment because the logic
does not account for the full set of potential candidates, especially in scenarios where multiple
daemons can reside on the same host.

Fixes: https://tracker.ceph.com/issues/73442
Signed-off-by: Shweta Bhosale <Shweta.Bhosale1@ibm.com>

ShwetaBhosale1 committed:
mgr/cephadm: Disable keepalived by default and start during serve loop if needed

If the ingress service is configured with a placement count lower than the number of hosts,
then when one node goes down, Keepalived is configured on the remaining hosts.
When the failed host comes back up, if it was previously configured as the master,
the VIP may become active on two hosts. To prevent this, Keepalived is disabled by default
and started within the service loop.

Fixes: https://tracker.ceph.com/issues/76304
Signed-off-by: Shweta Bhosale <Shweta.Bhosale1@ibm.com>

ShwetaBhosale1 committed:
mgr/cephadm: Scale down ingress service only when the backend NFS service is scaled down

When scaling ingress down with a ranked NFS backend that still has more
daemons than its placement target, remove ingress from hosts without NFS
first and skip removing co-located ingress on NFS hosts until the next
iteration after NFS scales down.

Fixes: https://tracker.ceph.com/issues/76304
Signed-off-by: Shweta Bhosale <Shweta.Bhosale1@ibm.com>

ShwetaBhosale1 committed:
mgr/cephadm: redeploy stopped keepalived when ingress deps change

Fixes: https://tracker.ceph.com/issues/76304
Signed-off-by: Shweta Bhosale <Shweta.Bhosale1@ibm.com>

ShwetaBhosale1 commented:
jenkins test make check

ShwetaBhosale1 commented:
jenkins test windows

ShwetaBhosale1 commented:
jenkins test make check

ShwetaBhosale1 commented:
jenkins test windows

rkachach commented:
Teuthology run was successful. Failing Test cases are not related to this PR:
https://pulpito.ceph.com/rkachach-2026-07-03_14:23:53-orch:cephadm-wip-rkachach-testing-2026-07-03-1453-distro-default-trial/
